### PR TITLE
Classification blog

### PR DESCRIPTION
**Pull request by aptsunny**
*Issue #, if available:*

fixes https://github.com/awslabs/autogluon/issues/265

*Description of changes:*
add dog classification script 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-234-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/1/index.html

File: autogluon/task/image_classification/image_classification.py [hunk 1]
```diff
@@ -33,13 +33,37 @@ def Dataset(*args, **kwargs):
            May either be a :class:`autogluon.task.image_classification.ImageFolderDataset`, :class:`autogluon.task.image_classification.RecordDataset`, 
            or a popular dataset already built into AutoGluon ('mnist', 'cifar10', 'cifar100', 'imagenet').
 
+        Parameters
+        ----------
+        name : str, optional
+            Which built-in dataset to use, will override all other options if specified.
+            The options are: 'mnist', 'cifar', 'cifar10', 'cifar100', 'imagenet'
+        traDataset for AutoGluon image classification tasks.
```
Comment (chongruo) [inline autogluon/task/image_classification/image_classification.py]: typo? (traDataset)

File: examples/image_classification/blog_test.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        momentum=ag.space.Real(0.90, 0.95),
+        wd=ag.space.Real(1e-6, 1e-4, log=True),
+        multi_precision=False
+    )
+    class NAG(optim.NAG):
+        pass
+
+    classifier = task.fit(dataset=load_dataset,
+                          # pretrain model
+                          # nets=ag.Categorical('resnext_101_standforddog', 'resnet151_v1d_standford'),
```
Comment (zhanghang1989) [inline examples/image_classification/blog_test.py]: `standford_dog_resnet152_v1` and `standford_dog_resnext101_64x4d` are in PR https://github.com/awslabs/autogluon/pull/236

**Comment by szha** [comment]
Job PR-234-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/4/index.html

**Comment by szha** [comment]
Job PR-234-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/6/index.html

**Comment by szha** [comment]
Job PR-234-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/8/index.html

File: autogluon/utils/file_helper.py [hunk 3]
```diff
@@ -138,6 +140,37 @@ def get_name(name):
...
         print('generate_csv B is done')
 
+def filter_value(prob, Threshold):
+    if prob > Threshold:
+        prob = prob
+    else:
+        prob = 0
+    return prob
+
+def generate_prob_csv(test_path, csv_path, preds, set_prob_thresh=0, custom='./submission.csv', scale_min_max=True, ensemble_list=''):
```
Comment (zhanghang1989) [inline autogluon/utils/file_helper.py]: This method need to be modified a little bit. The user flow/API should be 
```
test_dataset = task.Dataset('test', train=False)
inds, probs, probs_all = classifier.predict(test_dataset)
ag.utils.generate_prob_csv(test_dataset, probs_all, './submission.csv', set_prob_thresh=0.9
```

**Comment by szha** [comment]
Job PR-234-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/9/index.html

**Comment by szha** [comment]
Job PR-234-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/10/index.html

**Comment by szha** [comment]
Job PR-234-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/11/index.html

**Comment by szha** [comment]
Job PR-234-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/12/index.html

File: autogluon/task/image_classification/classifier.py [hunk 4]
```diff
@@ -112,37 +114,83 @@ def predict(self, X, input_size=224, plot=True):
...
-                transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
-            ])
-        def predict_img(img):
+        resize = int(math.ceil(input_size / resize_ratio))
+
+        transform_size = transforms.Compose([
+            transforms.Resize(resize),
+            transforms.CenterCrop(input_size),
+            transforms.ToTensor(),
+            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
```
Comment (larroy) [inline autogluon/task/image_classification/classifier.py]: why are hardcoded?

File: autogluon/task/image_classification/classifier.py [hunk 5]
```diff
@@ -112,37 +114,83 @@ def predict(self, X, input_size=224, plot=True):
...
-            return ind, probai, proba
+            if ensemble:
+                return proba
+            else:
+                ind = mx.nd.argmax(proba, axis=1).astype('int')
+                idx = mx.nd.stack(mx.nd.arange(proba.shape[0], ctx=proba.context), ind.astype('float32'))
+                probai = mx.nd.gather_nd(proba, idx)
+                return ind, probai, proba
+
+        def avg_prediction(different_dataset, threshold=0.001):
```
Comment (larroy) [inline autogluon/task/image_classification/classifier.py]: nontrivial function with docstring missing, could you please add to this and similar?

File: autogluon/task/image_classification/classifier.py [hunk 6]
```diff
@@ -112,37 +114,83 @@ def predict(self, X, input_size=224, plot=True):
...
+                proba_all = mx.nd.array(np.where(proba_all.asnumpy() >= threshold, proba_all.asnumpy(), 0))
+                ind = mx.nd.argmax(proba_all, axis=1).astype('int')
+                idx = mx.nd.stack(mx.nd.arange(proba_all.shape[0], ctx=proba_all.context), ind.astype('float32'))
+                proba = mx.nd.gather_nd(proba_all, idx)
+                inds.append(ind.asscalar())
+                probas.append(proba.asnumpy())
+                probals_all.append(proba_all.asnumpy().flatten())
+            return inds, probas, probals_all
+
+        def predict_imgs(X):
```
Comment (larroy) [inline autogluon/task/image_classification/classifier.py]: can you lowercase the argument (PEP8)  maybe with a more semantic name? Also docstring

**Comment by szha** [comment]
Job PR-234-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/13/index.html

File: autogluon/task/image_classification/classifier.py (hunk 4 above)
Comment (zhanghang1989) [reply autogluon/task/image_classification/classifier.py]: All the GluonCV model or even PyTorch pretrained models are using this normalization mean and std, as it is from ImageNet dataset. It is okay to hardcode, because we only do transfer learning for `ImageClassifcation.fit`

File: autogluon/task/image_classification/classifier.py (hunk 6 above)
Comment (aptsunny) [reply autogluon/task/image_classification/classifier.py]: > can you lowercase the argument (PEP8) maybe with a more semantic name? Also docstring

Thanks for your reminder i will add it immediately.

**Comment by zhanghang1989** [comment]
Hi @larroy , has the contributor addressed all your concerns?

**Comment by szha** [comment]
Job PR-234-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/15/index.html

**Comment by szha** [comment]
Job PR-234-16 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/16/index.html

**Comment by szha** [comment]
Job PR-234-17 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-234/17/index.html
